src/device-config.ts

```diff
@@ -0,0 +1,512 @@
+		};
+
+		for (const key in DeviceConfig.configKeys) {
+			const { envVarName } = DeviceConfig.configKeys[key];
+			const value = conf[key] != null ? (conf[key] as string) : '';
+			currentConf[envVarName] = value.toString();
+		}
+
+		return _.assign(currentConf, bootConfig);
+	}
+
+	public async formatConfigKeys(
+		conf: Dictionary<string>,
+	): Promise<Dictionary<any>> {
+		const backend = await this.getConfigBackend();
+		return await configUtils.formatConfigKeys(
+			backend,
+			DeviceConfig.validKeys,
+			conf,
+		);
+	}
+
+	public getDefaults() {
+		return _.extend(
+			{
+				SUPERVISOR_VPN_CONTROL: 'true',
+			},
+			_.mapValues(
+				_.mapKeys(DeviceConfig.configKeys, 'envVarName'),
+				'defaultValues',
+			),
+		);
+	}
+
+	private bootConfigChangeRequired(
+		configBackend: DeviceConfigBackend | null,
+		current: Dictionary<string>,
+		target: Dictionary<string>,
+	): boolean {
+		const targetBootConfig = configUtils.envToBootConfig(configBackend, target);
+		const currentBootConfig = configUtils.envToBootConfig(
+			configBackend,
+			current,
+		);
+
+		if (!_.isEqual(currentBootConfig, targetBootConfig)) {
+			_.each(targetBootConfig, (value, key) => {
+				// Ignore null check because we can't get here if configBackend is null
+				if (!configBackend!.isSupportedConfig(key)) {
+					if (currentBootConfig[key] !== value) {
+						const err = `Attempt to change blacklisted config value ${key}`;
+						this.logger.logSystemMessage(
+							err,
+							{ error: err },
+							'Apply boot config error',
+						);
+						throw new Error(err);
+					}
+				}
+			});
+			return true;
+		}
+		return false;
+	}
+
+	public async getRequiredSteps(
+		currentState: DeviceApplicationState,
+		targetState: DeviceApplicationState,
+	): Promise<ConfigStep[]> {
+		const current: Dictionary<string> = _.get(
+			currentState,
+			['local', 'config'],
+			{},
+		);
+		const target: Dictionary<string> = _.get(
+			targetState,
+			['local', 'config'],
+			{},
+		);
+
+		const steps: ConfigStep[] = [];
+
+		const unmanaged = await this.config.get('unmanaged');
+		const backend = await this.getConfigBackend();
+
+		const configChanges: Dictionary<string> = {};
+		const humanReadableConfigChanges: Dictionary<string> = {};
+		let reboot = false;
+
+		// If the legacy lock override is used, place it as the new variable
+		if (checkTruthy(target['OVERRIDE_LOCK'])) {
+			target['SUPERVISOR_OVERRIDE_LOCK'] = target['OVERRIDE_LOCK'];
+		}
+
+		_.each(
+			DeviceConfig.configKeys,
+			({ envVarName, varType, rebootRequired }, key) => {
+				// Test if the key is different
+				if (
+					!DeviceConfig.configTest(
+						varType,
+						current[envVarName],
+						target[envVarName],
+					)
+				) {
+					// Save the change if it is
+					configChanges[key] = target[envVarName];
+					humanReadableConfigChanges[envVarName] = target[envVarName];
+					reboot = reboot || (rebootRequired != null ? rebootRequired : false);
```

Comment:
I think `(rebootRequired != null ? rebootRequired : false)` can be simplified to just `rebootRequired` - if `rebootRequired == null` then it'll coerce to false anyway

Reply:
Nice